packages/service/ni_measurement_plugin_sdk_service/_internal/parameter/serialization_descriptors.py

```diff
@@ -0,0 +1,177 @@
+"""Serialization Descriptors."""
+
+from json import loads
+from typing import Any, List
+
+from google.protobuf import descriptor_pb2, descriptor_pool
+from google.protobuf.descriptor_pb2 import FieldDescriptorProto
+
+from ni_measurement_plugin_sdk_service._annotations import ENUM_VALUES_KEY
+from ni_measurement_plugin_sdk_service._internal.parameter.metadata import (
+    ParameterMetadata,
+)
+from tests.utilities.stubs.loopback.types_pb2 import ProtobufColor
+
+
+def _get_output_enum_type(
+    metadata_enum_list: List[str],
+    file_descriptor: descriptor_pb2.FileDescriptorProto,
+) -> Any:
+    """Get's matching enum class from 'file_descriptor'.
+
+    Args:
+        metadata_enum_list (List[str]): Enum names from metadata.annotations.
+
+        file_descriptor: Descriptor of proto file.
+
+    Returns:
+        Any: Matching enum class in a str type or None when enum is protobuf.
+    """
+    for enum_type in file_descriptor.enum_type:
+        enum_names = [enum_value.name for enum_value in enum_type.value]
+        if sorted(metadata_enum_list) == sorted(enum_names):
+            return enum_type.name
+    return None
+
+
+def _create_enum_type(
+    file_descriptor: descriptor_pb2.FileDescriptorProto,
+    parameter_metadata: ParameterMetadata,
+    field_descriptor: FieldDescriptorProto,
+) -> None:
+    """Implement a enum class in 'file_descriptor'.
+
+    Args:
+        file_descriptor (FileDescriptorProto): Descriptor of a proto file.
+
+        parmeter_metadata (ParameterMetadata): Metadata of current field.
+
+        field_descriptor (FieldDescriptorProto): Descriptor of a field.
+
+    Returns:
+        None: Only creates a enum class in file_descriptor.
+    """
+    enum_dict = loads(parameter_metadata.annotations[ENUM_VALUES_KEY])
+    if parameter_metadata.default_value is None:
+        enum_type_name = _get_output_enum_type(
+            metadata_enum_list=[enum for enum in enum_dict.keys()], file_descriptor=file_descriptor
+        )
+    else:
+        enum_type_name = _get_enum_type(parameter_metadata).__name__
+
+    if enum_type_name == "int" or enum_type_name is None:
+        field_descriptor.type_name = ProtobufColor.DESCRIPTOR.full_name
```

Comment:
This is where it references test code. Delete this or replace it with working code that doesn't rely on hardcoding details of the tests.

Reply:
Is it fine to create a enum type every time there is a enum passing in, regardless if it's a python or a protobuf enum? Since I'm not sure how to get the full_name from a protobuf enum which is why I was importing in the test code since it was defined in there.

If so, then what would be the name for the enum type when it's a protobuf? I've made it where it get's the first letter of each enum field name then joining them together, similar to a acronym without the dots but I don't think this is good enough.

Reply:
It is not fine to create new Python types every time you serialize or deserialize something.

If the service uses an enum, the Python type for the enum should already exist.

For example, this service uses the Function enum: https://github.com/ni/measurement-plugin-python/blob/main/examples/nidmm_measurement/measurement.py
Registering the service should not create another Python type for that enum. It can create a protobuf EnumDescriptor or whatever, though.

Reply:
Are you asking whether it's ok to create a new EnumDescriptor when the service is already using a protobuf enum? That might be ok.